### ISSUE TITLE
Stored Procedure and Migrations

### ISSUE THREAD

**Issue by agustinbcu01** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#776]
### Steps to reproduce
I create an empty migration on the UP Method, I write this.
```
  protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.Sql(@"
                   
                    DROP PROCEDURE IF EXISTS GetCurrentGeneration;
                       CREATE  PROCEDURE GetCurrentGeneration(IN timeOut INT, IN createBy char(255))
                        BEGIN
                                 DECLARE currentTimeStamp TIMESTAMP;
                                 DECLARE generationsCount INT;
                                 DECLARE sqlUpdated INT;
             
                                 SET currentTimeStamp = UTC_TIMESTAMP();

                        SET sqlUpdated = 1; -- SQL_SAFE_UPDATES;     
                        SET SQL_SAFE_UPDATES = 0;
                        UPDATE generetions 
                        SET 
                            status = 3
                        WHERE
                            DATE_ADD(UpdatedAt,
                                INTERVAL timeOut SECOND) < currentTimeStamp
                                AND Status IN (1 , 2);	
            
                                SET   SQL_SAFE_UPDATES = sqlUpdated;
                   

    			                    INSERT INTO generetions
                                    ( StartAt, Status, ImagesCount, VideosCount, CreatedBy, CreatedAt, UpdatedBy, UpdatedAt)
                                    SELECT * FROM (SELECT currentTimeStamp, 1 as a, 0 as b, 0 as c, createBy as d, currentTimeStamp as e, createBy as f,currentTimeStamp as g) AS tmp
                                    WHERE NOT EXISTS (SELECT id from generetions WHERE   Status IN (1 , 2)) limit 1;
         
              
                        SELECT 
                            *
                        FROM
                            generetions
                        WHERE
                            Status IN (1 , 2)
                        ORDER BY UpdatedAt DESC;
        
                   
         
                        END
            ");
        }
```
and I generate the migration sql script.
```

DROP PROCEDURE IF EXISTS MigrationsScript;
DELIMITER //
CREATE PROCEDURE MigrationsScript()
BEGIN
    IF NOT EXISTS(SELECT 1 FROM `__EFMigrationsHistory` WHERE `MigrationId` = '20190411134055_AddGenerations_sp') THEN
    
                       
                       DROP PROCEDURE IF EXISTS GetCurrentGeneration;
                       CREATE  PROCEDURE GetCurrentGeneration(IN timeOut INT, IN createBy char(255))
                        BEGIN
                                 DECLARE currentTimeStamp TIMESTAMP;
                                 DECLARE generationsCount INT;
                                 DECLARE sqlUpdated INT;
             
                                 SET currentTimeStamp = UTC_TIMESTAMP();

                        SET sqlUpdated = 1; -- SQL_SAFE_UPDATES;     
                        SET SQL_SAFE_UPDATES = 0;
                        UPDATE generetions 
                        SET 
                            status = 3
                        WHERE
                            DATE_ADD(UpdatedAt,
                                INTERVAL timeOut SECOND) < currentTimeStamp
                                AND Status IN (1 , 2);	
            
                                SET   SQL_SAFE_UPDATES = sqlUpdated;
                   

    			                    INSERT INTO generetions
                                    ( StartAt, Status, ImagesCount, VideosCount, CreatedBy, CreatedAt, UpdatedBy, UpdatedAt)
                                    SELECT * FROM (SELECT currentTimeStamp, 1 as a, 0 as b, 0 as c, createBy as d, currentTimeStamp as e, createBy as f,currentTimeStamp as g) AS tmp
                                    WHERE NOT EXISTS (SELECT id from generetions WHERE   Status IN (1 , 2)) limit 1;
         
              
                        SELECT 
                            *
                        FROM
                            generetions
                        WHERE
                            Status IN (1 , 2)
                        ORDER BY UpdatedAt DESC;
        
                   
         
                        END;
					
				
				
END IF;
END //
DELIMITER ;
CALL MigrationsScript();
DROP PROCEDURE MigrationsScript;


DROP PROCEDURE IF EXISTS MigrationsScript;
DELIMITER //
CREATE PROCEDURE MigrationsScript()
BEGIN
    IF NOT EXISTS(SELECT 1 FROM `__EFMigrationsHistory` WHERE `MigrationId` = '20190411134055_AddGenerations_sp') THEN

    INSERT INTO `__EFMigrationsHistory` (`MigrationId`, `ProductVersion`)
    VALUES ('20190411134055_AddGenerations_sp', '2.2.4-servicing-10062');

    END IF;
END //
DELIMITER ;
CALL MigrationsScript();
DROP PROCEDURE MigrationsScript;

 
```
### The issue
When the migration script is run rerturn this error: ` Error Code: 1357. Can't drop or alter a PROCEDURE from within another stored routine`

MySQL version: 8.0.11
Operating system: Windows 10
Pomelo.EntityFrameworkCore.MySql version: 2.2.0
Microsoft.AspNetCore.App version: 2.2.4

**Comment by mguinness** [comment]
@Meberem This was introduced in https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/pull/683, do you have any suggestions/workarounds?

**Comment by Meberem** [comment]
Kind of tricky as this is user generated. If it was from code gen you would be able to tell that it doesn’t need the wrapper and not write it out. As it stands, my gut reaction would be to have an overload or a `.StoredProcedure()` and have some documentation detailing how and why. I’m not familiar enough with MySQL to know if there is a better way of doing this, sorry

**Comment by agustinbcu01** [comment]
@mguinness  I run out ideas. I was playing with the migration SQL and have not look.

**Comment by souchprod** [comment]
Hi, do you have a workaround for this ?
Issue closed but still a blocking one for me :'(

**Comment by agustinbcu01** [comment]
It is MySql limitation.

**Comment by souchprod** [comment]
Hum, as it is implemented today yes, if trigger an engine limitation but if we can just extract the POMELO_BEFORE_DROP_PRIMARY_KEY procedure creation and removal to outside of the migration script itself, we could fix it.

Currently generated statement :

```mysql
DROP PROCEDURE IF EXISTS MigrationsScript;
DELIMITER //
CREATE PROCEDURE MigrationsScript()
BEGIN
    IF NOT EXISTS(SELECT 1 FROM `__EFMigrationsHistory` WHERE `MigrationId` = 'migration_name_here') THEN

    DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;
    CREATE PROCEDURE POMELO_BEFORE_DROP_PRIMARY_KEY(IN `SCHEMA_NAME_ARGUMENT` VARCHAR(255), IN `TABLE_NAME_ARGUMENT` VARCHAR(255))
    BEGIN
    	DECLARE HAS_AUTO_INCREMENT_ID TINYINT(1);
    	DECLARE PRIMARY_KEY_COLUMN_NAME VARCHAR(255);
    	DECLARE PRIMARY_KEY_TYPE VARCHAR(255);
    	DECLARE SQL_EXP VARCHAR(1000);
    	SELECT COUNT(*)
    		INTO HAS_AUTO_INCREMENT_ID
    		FROM `information_schema`.`COLUMNS`
    		WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
    			AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
    			AND `Extra` = 'auto_increment'
    			AND `COLUMN_KEY` = 'PRI'
    			LIMIT 1;
    	IF HAS_AUTO_INCREMENT_ID THEN
    		SELECT `COLUMN_TYPE`
    			INTO PRIMARY_KEY_TYPE
    			FROM `information_schema`.`COLUMNS`
    			WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
    				AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
    				AND `COLUMN_KEY` = 'PRI'
    			LIMIT 1;
    		SELECT `COLUMN_NAME`
    			INTO PRIMARY_KEY_COLUMN_NAME
    			FROM `information_schema`.`COLUMNS`
    			WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
    				AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
    				AND `COLUMN_KEY` = 'PRI'
    			LIMIT 1;
    		SET SQL_EXP = CONCAT('ALTER TABLE `', (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA())), '`.`', TABLE_NAME_ARGUMENT, '` MODIFY COLUMN `', PRIMARY_KEY_COLUMN_NAME, '` ', PRIMARY_KEY_TYPE, ' NOT NULL;');
    		SET @SQL_EXP = SQL_EXP;
    		PREPARE SQL_EXP_EXECUTE FROM @SQL_EXP;
    		EXECUTE SQL_EXP_EXECUTE;
    		DEALLOCATE PREPARE SQL_EXP_EXECUTE;
    	END IF;
    END;
    CALL POMELO_BEFORE_DROP_PRIMARY_KEY(NULL, 'table_name_here');
    DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;
    ALTER TABLE `table_name_here` DROP PRIMARY KEY;

    END IF;
END //
DELIMITER ;
CALL MigrationsScript();
DROP PROCEDURE MigrationsScript;
```
Fixed statement:

```mysql
DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;

DELIMITER //

CREATE PROCEDURE POMELO_BEFORE_DROP_PRIMARY_KEY(IN `SCHEMA_NAME_ARGUMENT` VARCHAR(255), IN `TABLE_NAME_ARGUMENT` VARCHAR(255))
BEGIN
	DECLARE HAS_AUTO_INCREMENT_ID TINYINT(1);
	DECLARE PRIMARY_KEY_COLUMN_NAME VARCHAR(255);
	DECLARE PRIMARY_KEY_TYPE VARCHAR(255);
	DECLARE SQL_EXP VARCHAR(1000);
	SELECT COUNT(*)
		INTO HAS_AUTO_INCREMENT_ID
		FROM `information_schema`.`COLUMNS`
		WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
			AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
			AND `Extra` = 'auto_increment'
			AND `COLUMN_KEY` = 'PRI'
			LIMIT 1;
	IF HAS_AUTO_INCREMENT_ID THEN
		SELECT `COLUMN_TYPE`
			INTO PRIMARY_KEY_TYPE
			FROM `information_schema`.`COLUMNS`
			WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
				AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
				AND `COLUMN_KEY` = 'PRI'
			LIMIT 1;
		SELECT `COLUMN_NAME`
			INTO PRIMARY_KEY_COLUMN_NAME
			FROM `information_schema`.`COLUMNS`
			WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
				AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
				AND `COLUMN_KEY` = 'PRI'
			LIMIT 1;
		SET SQL_EXP = CONCAT('ALTER TABLE `', (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA())), '`.`', TABLE_NAME_ARGUMENT, '` MODIFY COLUMN `', PRIMARY_KEY_COLUMN_NAME, '` ', PRIMARY_KEY_TYPE, ' NOT NULL;');
		SET @SQL_EXP = SQL_EXP;
		PREPARE SQL_EXP_EXECUTE FROM @SQL_EXP;
		EXECUTE SQL_EXP_EXECUTE;
		DEALLOCATE PREPARE SQL_EXP_EXECUTE;
	END IF;
END //

DELIMITER ;

DROP PROCEDURE IF EXISTS MigrationsScript;
DELIMITER //
CREATE PROCEDURE MigrationsScript()
BEGIN

  IF NOT EXISTS(SELECT 1 FROM `__EFMigrationsHistory` WHERE `MigrationId` = 'migration_name_here') THEN  
 
    CALL POMELO_BEFORE_DROP_PRIMARY_KEY(NULL, 'table_name_here');
    ALTER TABLE `table_name_here` DROP PRIMARY KEY;

  END IF;

END //
DELIMITER ;
CALL MigrationsScript();
DROP PROCEDURE MigrationsScript;

DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;
```

So, would need to always create in the beginning of the script & drop at the end this kind of procedure (only if required by one of the script would be better).

**Comment by igrmk** [comment]
Isn't this issue a showstopper?
It leads to a bad idempotent migration script on just dropping primary key.
So you cannot even rename a table.

Maybe it is a stupid question...
But why to wrap migrations into procedures at all?

BTW such nonidempotent migrations don't work as well currently.
But this is because of delimiters.
I'll report it later.

**Comment by lauxjpn** [comment]
> But why to wrap mMaybe it is a stupid question...
But why to wrap migrations into procedures at all?

@igrmk Not a stupid question at all. This needs to be rechecked. If a SP is not necessary here, it should be implemented without one.

**Comment by lauxjpn** [comment]
@agustinbcu01 As you found out yourself, there is currently no reliable way to create a stored procedure within a stored procedure.
Because idempotent script generation requires the use of control flow statements (`IF`) and those can only be used in stored procedures, creating stored procedures inside idempotent migrations scripts is impossible at the moment (@igrmk). And because `PREPARE` statements do not support `CREATE PROCEDURE`, these cannot be used as a workaround either.
The only workaround I know of is described on [StackOverflow](https://stackoverflow.com/a/50671306/2618319), but requires`INSERT` permissions for the `mysql.proc` table.
So the only option I can see here for a semi-automated approach, is to take the script from https://github.com/aspnet/EntityFrameworkCore/issues/10717#issuecomment-358462348 to split up the migrations in individual files and then run some logic and script before or each applied migration.

Non-idempotent migrations should work just fine.

**Comment by agustinbcu01** [comment]
Thanks for look at his issue.

The file approach looks good, now I'm using a custom SQL call in my project.
I'll test the file idea and compare store procedure Vs the custom script.

Regards.

**Comment by Asier-Villanueva** [comment]
A solution could be to replace the internal procedure with inline statements.

Currently generated statement:



``` mysql
DROP PROCEDURE IF EXISTS MigrationsScript;
DELIMITER //
CREATE PROCEDURE MigrationsScript()
BEGIN
    IF NOT EXISTS(SELECT 1 FROM `__EFMigrationsHistory` WHERE `MigrationId` = 'migration_name_here') THEN

    DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;
    CREATE PROCEDURE POMELO_BEFORE_DROP_PRIMARY_KEY(IN `SCHEMA_NAME_ARGUMENT` VARCHAR(255), IN `TABLE_NAME_ARGUMENT` VARCHAR(255))
    BEGIN
    	DECLARE HAS_AUTO_INCREMENT_ID TINYINT(1);
    	DECLARE PRIMARY_KEY_COLUMN_NAME VARCHAR(255);
    	DECLARE PRIMARY_KEY_TYPE VARCHAR(255);
    	DECLARE SQL_EXP VARCHAR(1000);
    	SELECT COUNT(*)
    		INTO HAS_AUTO_INCREMENT_ID
    		FROM `information_schema`.`COLUMNS`
    		WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
    			AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
    			AND `Extra` = 'auto_increment'
    			AND `COLUMN_KEY` = 'PRI'
    			LIMIT 1;
    	IF HAS_AUTO_INCREMENT_ID THEN
    		SELECT `COLUMN_TYPE`
    			INTO PRIMARY_KEY_TYPE
    			FROM `information_schema`.`COLUMNS`
    			WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
    				AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
    				AND `COLUMN_KEY` = 'PRI'
    			LIMIT 1;
    		SELECT `COLUMN_NAME`
    			INTO PRIMARY_KEY_COLUMN_NAME
    			FROM `information_schema`.`COLUMNS`
    			WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
    				AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
    				AND `COLUMN_KEY` = 'PRI'
    			LIMIT 1;
    		SET SQL_EXP = CONCAT('ALTER TABLE `', (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA())), '`.`', TABLE_NAME_ARGUMENT, '` MODIFY COLUMN `', PRIMARY_KEY_COLUMN_NAME, '` ', PRIMARY_KEY_TYPE, ' NOT NULL;');
    		SET @SQL_EXP = SQL_EXP;
    		PREPARE SQL_EXP_EXECUTE FROM @SQL_EXP;
    		EXECUTE SQL_EXP_EXECUTE;
    		DEALLOCATE PREPARE SQL_EXP_EXECUTE;
    	END IF;
    END;
    CALL POMELO_BEFORE_DROP_PRIMARY_KEY(NULL, 'table_name_here');
    DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;
    ALTER TABLE `table_name_here` DROP PRIMARY KEY;

    END IF;
END //
DELIMITER ;
CALL MigrationsScript();
DROP PROCEDURE MigrationsScript;
```

Fixed statement:

``` mysql

DROP PROCEDURE IF EXISTS MigrationsScript;
DELIMITER //
CREATE PROCEDURE MigrationsScript()
BEGIN
    DECLARE HAS_AUTO_INCREMENT_ID TINYINT(1);
    DECLARE PRIMARY_KEY_COLUMN_NAME VARCHAR(255);
    DECLARE PRIMARY_KEY_TYPE VARCHAR(255);
    DECLARE SQL_EXP VARCHAR(1000);
    DECLARE SCHEMA_NAME_ARGUMENT VARCHAR(255);
    DECLARE TABLE_NAME_ARGUMENT VARCHAR(255);

    IF NOT EXISTS(SELECT 1 FROM `__EFMigrationsHistory` WHERE `MigrationId` = 'migration_name_here') THEN
    
        SET SCHEMA_NAME_ARGUMENT = NULL;
        SET TABLE_NAME_ARGUMENT = 'table_name_here';

    	SELECT COUNT(*)
    		INTO HAS_AUTO_INCREMENT_ID
    		FROM `information_schema`.`COLUMNS`
    		WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
    			AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
    			AND `Extra` = 'auto_increment'
    			AND `COLUMN_KEY` = 'PRI'
    			LIMIT 1;
    	IF HAS_AUTO_INCREMENT_ID THEN
    		SELECT `COLUMN_TYPE`
    			INTO PRIMARY_KEY_TYPE
    			FROM `information_schema`.`COLUMNS`
    			WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
    				AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
    				AND `COLUMN_KEY` = 'PRI'
    			LIMIT 1;
    		SELECT `COLUMN_NAME`
    			INTO PRIMARY_KEY_COLUMN_NAME
    			FROM `information_schema`.`COLUMNS`
    			WHERE `TABLE_SCHEMA` = (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA()))
    				AND `TABLE_NAME` = TABLE_NAME_ARGUMENT
    				AND `COLUMN_KEY` = 'PRI'
    			LIMIT 1;
    		SET SQL_EXP = CONCAT('ALTER TABLE `', (SELECT IFNULL(SCHEMA_NAME_ARGUMENT, SCHEMA())), '`.`', TABLE_NAME_ARGUMENT, '` MODIFY COLUMN `', PRIMARY_KEY_COLUMN_NAME, '` ', PRIMARY_KEY_TYPE, ' NOT NULL;');
    		SET @SQL_EXP = SQL_EXP;
    		PREPARE SQL_EXP_EXECUTE FROM @SQL_EXP;
    		EXECUTE SQL_EXP_EXECUTE;
    		DEALLOCATE PREPARE SQL_EXP_EXECUTE;
    	END IF;

    ALTER TABLE `table_name_here` DROP PRIMARY KEY;

    END IF;
END //
DELIMITER ;
CALL MigrationsScript();
DROP PROCEDURE MigrationsScript;

```

**Comment by lauxjpn** [comment]
@Asier-Villanueva Please elaborate on what this fixes and why.

As far as I can see, you just removed anything related to the `POMELO_BEFORE_DROP_PRIMARY_KEY()` call without any substitution, which can lead to errors when executing `DropPrimaryKeyOperation` migration operations.

Also, its not the Pomelo internal procedures that are the issue (at least they are supposed to work flawlessly in releases since about 3.0.0), but custom procedures that users define themselves.

**Comment by Asier-Villanueva** [comment]
@lauxjpn Ok. I'm still using version 2.2.6. We're still migrating our apps to ef core 3. 
We'll have to accelerate migration.

**Comment by lauxjpn** [comment]
@Asier-Villanueva You can also just use the non-idempotent script generation, which should work in this case, or just manually add the stored procedures once to the beginning of the script, remove them at the end of the script and just keep the calls in between.

> We'll have to accelerate migration.

EF Core 5.0 indroduces `.AsSplitQuery()`, which can give you query execution behavior similar to 2.2.6. We have not released a compatible provider yet, but it won't be much longer.

**Comment by Asier-Villanueva** [comment]
@lauxjpn I need idempotent script for some automated tools I'm using in deployments. For the moment I've cretated custom implementations of IMigrator and IMigrationsSqlGenerator to create procedures at the beginning of the script and to override the script generated for Add Primary Key and Drop Primary Key operations.

``` c#

    public class CustomMigrator : Migrator
    {
        private const string BEFORE_DROP_PRIMARY_KEY_PROCEDURE_ADD = @"DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;
                CREATE PROCEDURE POMELO_BEFORE_DROP_PRIMARY_KEY(IN `SCHEMA_NAME_ARGUMENT` VARCHAR(255), IN `TABLE_NAME_ARGUMENT` VARCHAR(255))
                BEGIN
                .....
                END;";
        private const string BEFORE_DROP_PRIMARY_KEY_PROCEDURE_REMOVE = @"DROP PROCEDURE IF EXISTS POMELO_BEFORE_DROP_PRIMARY_KEY;";

        private const string AFTER_ADD_PRIMARY_KEY_PROCEDURE_ADD = @"DROP PROCEDURE IF EXISTS POMELO_AFTER_ADD_PRIMARY_KEY;
                CREATE PROCEDURE POMELO_AFTER_ADD_PRIMARY_KEY(IN `SCHEMA_NAME_ARGUMENT` VARCHAR(255), IN `TABLE_NAME_ARGUMENT` VARCHAR(255), IN `COLUMN_NAME_ARGUMENT` VARCHAR(255))
                BEGIN
                ...
                END;";
        private const string AFTER_ADD_PRIMARY_KEY_REMOVE = "DROP PROCEDURE IF EXISTS POMELO_AFTER_ADD_PRIMARY_KEY;";

        public CustomMigrator([NotNullAttribute] IMigrationsAssembly migrationsAssembly, [NotNullAttribute] IHistoryRepository historyRepository, [NotNullAttribute] IDatabaseCreator databaseCreator, [NotNullAttribute] IMigrationsSqlGenerator migrationsSqlGenerator, [NotNullAttribute] IRawSqlCommandBuilder rawSqlCommandBuilder, [NotNullAttribute] IMigrationCommandExecutor migrationCommandExecutor, [NotNullAttribute] IRelationalConnection connection, [NotNullAttribute] ISqlGenerationHelper sqlGenerationHelper, [NotNullAttribute] IDiagnosticsLogger<DbLoggerCategory.Migrations> logger, [NotNullAttribute] IDatabaseProvider databaseProvider) 
            : base(migrationsAssembly, historyRepository, databaseCreator, migrationsSqlGenerator, rawSqlCommandBuilder, migrationCommandExecutor, connection, sqlGenerationHelper, logger, databaseProvider) { }

        public override string GenerateScript(string fromMigration = null, string toMigration = null, bool idempotent = false)
        {
            var builder = new StringBuilder();
            builder.AppendLine(AFTER_ADD_PRIMARY_KEY_PROCEDURE_ADD);
            builder.AppendLine(BEFORE_DROP_PRIMARY_KEY_PROCEDURE_ADD);

            builder.Append(base.GenerateScript(fromMigration, toMigration, idempotent));

            builder.AppendLine(AFTER_ADD_PRIMARY_KEY_REMOVE);
            builder.AppendLine(BEFORE_DROP_PRIMARY_KEY_PROCEDURE_REMOVE);

            return builder.ToString();
        }
    }


    public class CustomMigrationsSqlGenerator : MySqlMigrationsSqlGenerator
    {
        public CustomMigrationsSqlGenerator([NotNullAttribute] MigrationsSqlGeneratorDependencies dependencies, [NotNullAttribute] IMigrationsAnnotationProvider migrationsAnnotations, [NotNullAttribute] IMySqlOptions options) 
            : base(dependencies, migrationsAnnotations, options) { }

        protected override void Generate(AddPrimaryKeyOperation operation, IModel model, MigrationCommandListBuilder builder)
        {
            if (operation is null) { throw new ArgumentNullException(nameof(operation)); }
            if (builder is null) { throw new ArgumentNullException(nameof(builder)); }

            builder
                .Append("ALTER TABLE ")
                .Append(Dependencies.SqlGenerationHelper.DelimitIdentifier(operation.Table, operation.Schema))
                .Append(" ADD ");
            PrimaryKeyConstraint(operation, model, builder);
            builder.AppendLine(Dependencies.SqlGenerationHelper.StatementTerminator);

            if (operation.Columns.Length == 1)
            {
                if (operation.Schema == null)
                {
                    builder.Append(
                        $"CALL POMELO_AFTER_ADD_PRIMARY_KEY(NULL, '{operation.Table}', '{operation.Columns.First()}');");
                }
                else
                {
                    builder.Append(
                        $"CALL POMELO_AFTER_ADD_PRIMARY_KEY('{operation.Schema}', '{operation.Table}', '{operation.Columns.First()}');");
                }

                builder.AppendLine();
            }

            EndStatement(builder);
        }

        protected override void Generate(DropPrimaryKeyOperation operation, IModel model, MigrationCommandListBuilder builder)
        {
            if (operation is null) { throw new ArgumentNullException(nameof(operation)); }
            if (builder is null) { throw new ArgumentNullException(nameof(builder)); }

            if (string.IsNullOrWhiteSpace(operation.Schema))
            {
                builder.Append($"CALL POMELO_BEFORE_DROP_PRIMARY_KEY(NULL, '{operation.Table}');");
            }
            else
            {
                builder.Append($"CALL POMELO_BEFORE_DROP_PRIMARY_KEY('{operation.Schema}', '{operation.Table}');");
            }

            builder.AppendLine();

            builder
                .Append("ALTER TABLE ")
                .Append(Dependencies.SqlGenerationHelper.DelimitIdentifier(operation.Table, operation.Schema))
                .Append(" DROP PRIMARY KEY;")
                .AppendLine();

            EndStatement(builder);
        }
    }
````

**Comment by amura11** [comment]
In case this can help someone in the future, I ran into this issue with 5.0.0 and had a hard time figuring out if it had been fixed or not. Eventually I stumbled across another conversation that confirmed this issue is fixed in 5.0.3 and 6.0.0. 

I can confirm updating to 5.0.3 now generates the procedures at the start of the script and no longer causes this issue.
